File: app/src/main/java/com/inter/trunks/base/component/RecyclerUIComponent.java
package com.inter.trunks.base.component;

import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import java.util.HashSet;
import java.util.Iterator;

public class RecyclerUIComponent<Adapter extends RecyclerView.Adapter> extends UIComponent {
    protected RecyclerView recyclerView;
    @Nullable
    protected View emptyView;

    private Adapter adapter;

    public RecyclerUIComponent(Adapter adapter) {
        this.adapter = adapter;
    }

    private RecyclerView.LayoutManager layoutManager;
    private HashSet<RecyclerView.ItemDecoration> setItemDecorations = new HashSet();

    @Override
    public void onViewCreated(View layout) {
        super.onViewCreated(layout);
        recyclerView = layout.findViewById(R.id.list);
        emptyView = layout.findViewById(android.R.id.empty);
        if (recyclerView.getLayoutManager() == null) {
            if (layoutManager == null) {
                layoutManager = new LinearLayoutManager(getContext());
            }
            recyclerView.setLayoutManager(layoutManager);
        }
        recyclerView.setAdapter(adapter);
        if (setItemDecorations.size() > 0) {
            Iterator<RecyclerView.ItemDecoration> itr = setItemDecorations.iterator();
            while (itr.hasNext()) {
                recyclerView.addItemDecoration(itr.next());
            }
        }
    }

    public void setLayoutManager(RecyclerView.LayoutManager layoutManager) {
        this.layoutManager = layoutManager;
    }

    public void refreshUI() {
        if (adapter.getItemCount() == 0) {
            showEmptyView();
        } else {
            hideEmptyView();
        }
    }

    public void addItemDecorations(HashSet<RecyclerView.ItemDecoration> setItemDecorations) {
        this.setItemDecorations.addAll(setItemDecorations);
    }

    public void showEmptyView() {
        recyclerView.setVisibility(View.GONE);
        if (emptyView != null) emptyView.setVisibility(View.VISIBLE);
    }

    public void hideEmptyView() {
        recyclerView.setVisibility(View.VISIBLE);
        if (emptyView != null) emptyView.setVisibility(View.GONE);
    }

    public void hideViews() {
        recyclerView.setVisibility(View.GONE);
        if (emptyView != null) emptyView.setVisibility(View.GONE);
    }

    public Adapter getAdapter() {
        return adapter;
    }

    public RecyclerView getRecyclerView() {
        return recyclerView;
    }

    public void setEmptyText(@StringRes int text) {
        setEmptyText(getContext().getString(text));
    }

    public void setEmptyText(String text) {
        if (emptyView instanceof TextView) {
            ((TextView) emptyView).setText(text);
        } else throw new IllegalStateException("EmptyView must be an instance of TextView");
    }

    @Override
    public void onDestroyView() {
        recyclerView.setAdapter(null);
        recyclerView.setLayoutManager(null);
        recyclerView = null;
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        layoutManager = null;
        setItemDecorations.clear();
        adapter = null;
        super.onDestroy();
    }
}

File: app/src/main/java/com/inter/trunks/base/component/RecyclerUIComponent.kt
package ru.mos.polls.kbase.component

import android.support.annotation.StringRes
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.TextView
import org.jetbrains.anko.findOptional
import org.koin.core.parameter.parametersOf
import org.koin.standalone.KoinComponent
import org.koin.standalone.get
import ru.mos.polls.R
import ru.mos.polls.base.component.UIComponent
import ru.mos.polls.kbase.utils.show

/**
 * Компонент для работы со списком RecyclerView.
 *
 * RecyclerView - R.id.list.
 * EmptyView - android.R.id.empty. Будет показываться, если список пуст.
 *
 */
class RecyclerUIComponent<Adapter : RecyclerView.Adapter<*>>(

        var adapter: Adapter? = null

) : UIComponent(), KoinComponent {

    var recycler: RecyclerView? = null
        private set

    /**
     * Отображается View-в случае пустого результата.
     */
    private var emptyView: View? = null

    var recyclerItemDecoration: MutableSet<RecyclerView.ItemDecoration> = mutableSetOf()

    var layoutManager: RecyclerView.LayoutManager? = null

    override fun onViewCreated(layout: View?) {
        super.onViewCreated(layout)
        layout?.let { view ->
            recycler = view.findOptional(R.id.list)
            emptyView = view.findOptional(android.R.id.empty)
        }
        emptyView?.show(false)
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        recycler?.let { list ->
            with(list) {
                if (this@RecyclerUIComponent.layoutManager == null)
                    this@RecyclerUIComponent.layoutManager =
                            get<LinearLayoutManager> { parametersOf(context) }

                layoutManager = this@RecyclerUIComponent.layoutManager

                adapter = this@RecyclerUIComponent.adapter

                recyclerItemDecoration.forEach { addItemDecoration(it) }
            }
        }
    }

    fun refreshUI() {
        adapter?.let { adapter ->
            if (adapter.itemCount > 0)
                hideEmptyView()
            else
                showEmptyView()
        } ?: showEmptyView()
    }

    private fun showEmptyView() {
        recycler?.show(false)
        emptyView?.show(true)
    }

    private fun hideEmptyView() {
        recycler?.show(true)
        emptyView?.show(false)
    }

    fun hideViews() {
        recycler?.show(false)
        emptyView?.show(false)
    }

    override fun onDestroy() {
        clear()
        recyclerItemDecoration.clear()
        adapter = null
        layoutManager = null
        super.onDestroy()
    }

    override fun onDestroyView() {
        clear()
        super.onDestroyView()
    }

    private fun clear() {
        recycler?.let {
            it.adapter = null
            it.layoutManager = null
        }
        recycler = null
    }

    fun setEmptyViewText(@StringRes stringId: Int) {
        if (emptyView is TextView) {
            (emptyView as TextView).text = context.getString(stringId)
        }
    }

    fun scrollToPosition(position: Int) = recycler?.scrollToPosition(position)

}
